src/main/java/com/google/devtools/build/lib/runtime/QueryRuntimeHelper.java

```diff
@@ -63,8 +69,31 @@ class StdoutQueryRuntimeHelperFactory implements Factory {
     private StdoutQueryRuntimeHelperFactory() {}
 
     @Override
-    public QueryRuntimeHelper create(CommandEnvironment env) {
-      return createInternal(env.getReporter().getOutErr().getOutputStream());
+    public QueryRuntimeHelper create(CommandEnvironment env, CommonQueryOptions options)
+        throws QueryRuntimeHelperException {
+      if (Strings.isNullOrEmpty(options.outputFile)) {
+        return createInternal(env.getReporter().getOutErr().getOutputStream());
+      } else {
+        InstrumentationOutput instrumentationOutput =
+            env.getRuntime()
+                .getInstrumentationOutputFactory()
+                .createInstrumentationOutput(
+                    "query_output",
+                    PathFragment.create(options.outputFile),
+                    DestinationRelativeTo.WORKING_DIR,
+                    env,
+                    env.getReporter(),
+                    false,
+                    false);
+        try {
+          return new FileQueryRuntimeHelper(instrumentationOutput);
+        } catch (IOException e) {
+          throw new QueryRuntimeHelperException(
+              "Could not open query output file " + instrumentationOutput.getHumanReadableName(),
```

Comment:
i think you alluded to this in my other comment - is this the absolute path? a perhaps not-well-documented thing we like to try to do is avoid absolute paths in user-visible messaging as much as possible, this way if the local filesystem (CI etc) doesn't match the end-user's filesystem there's less confusion. also, as you may have deduced from the instrumentation-output abstraction, this path might not reflect the actual file we were operating on.

that said, logging absolute paths to info logging should be fine.

Reply:
Hmm, okay. Personally when Bazel outputs a path, it can be difficult to understand which of Bazel's many directories it is relative to. So I would prefer Bazel only print absolute paths :)